snoop/walker.py

```diff
@@ -7,70 +7,78 @@
 
 class Walker(object):
 
-    def __init__(self, generation, root, prefix, restart):
-        self.generation = generation
+    def __init__(self, root, prefix, restart, container_doc):
         self.root = Path(root)
         self.prefix = Path(prefix) if prefix else None
+        self.container_doc = container_doc
+        self.documents = []
         if restart:
             models.FolderMark.objects.all().delete()
 
-
     @classmethod
     def walk(cls, *args):
         self = cls(*args)
         try:
-            return self.handle(self.root / self.prefix if self.prefix else None)
+            first_item = self.root / self.prefix if self.prefix else self.root
+            first_parent = self.container_doc
+            self.handle(first_item, first_parent)
         except KeyboardInterrupt:
             pass
+        return self.documents
 
     def _path(self, file):
         return file.relative_to(self.root)
 
-    def handle(self, item=None):
-        if item is None:
-            item = self.root
-
+    def handle(self, item, parent):
         if item.is_dir():
-            self.handle_folder(item)
-
+            self.handle_folder(item, parent)
         else:
-            self.handle_file(item)
+            self.handle_file(item, parent)
 
-    def handle_folder(self, folder):
+    def handle_folder(self, folder, parent):
         path = self._path(folder)
-        print('FOLDER', path)
-        if models.FolderMark.objects.filter(path=path).count():
-            print('SKIP', path)
-            return
-        if str(path) != '.':
-            models.Document.objects.get_or_create(
+        if str(path) == '.':
+            if self.container_doc:
+                new_doc = self.container_doc
+            else:
+                new_doc, _ = models.Document.objects.get_or_create(
+                    path='',
+                    disk_size=0,
+                    content_type=FOLDER,
+                    filename='',
+                )
+        else:
+            new_doc, created = models.Document.objects.get_or_create(
                 path=path,
                 disk_size=0,
                 content_type=FOLDER,
                 filename=path.name,
+                parent=parent,
+                container=self.container_doc,
             )
+            self.documents.append((new_doc, created))
         for child in folder.iterdir():
-            self.handle(child)
-        models.FolderMark.objects.create(path=path)
```

Comment:
Good, we should have removed `FolderMark` a long time ago, let's also remove the model.